kernel/src/process/signal/mod.rs

```diff
@@ -69,6 +70,18 @@ pub fn handle_pending_signal(user_ctx: &mut UserContext, ctx: &Context) -> Resul
             restorer_addr,
             mask,
         } => {
+            if user_ctx.syscall_num() != usize::MAX
+                && user_ctx.syscall_ret() == -(Errno::ERESTARTSYS as i32) as usize
```

Comment:
**Major issue 1.** We cannot check the conditions this way. The branch gated by these conditions are _only_ for system call handling.  But `handle_pending_signal` may be called after either system call or exception handling. If it is not a system call, then these checks do not make any sense.

Reply:
Just to put it clear, I added some logics in the assembly and signal-related code so that the rax would be -1 if it's not coming from a syscall. After all, I agree that we should not add more assembly logics unless necessary so I have modified as you advised.